tests/EdgeToEdge/Routes/UpdateDonorRouteTest.php

```diff
@@ -0,0 +1,265 @@
+				$factory->getEntityManager()->flush();
+
+				$this->performRequest(
+					$client,
+					$this->newPrivateDonorData(),
+					$donation->getId(),
+					self::CORRECT_UPDATE_TOKEN,
+					self::CORRECT_UPDATE_TOKEN
+				);
+
+				$response = $client->getResponse();
+				$this->assertTrue( $response->isForbidden() );
+			}
+		);
+	}
+
+	public function testWhenDonationDataIsInvalid_requestIsDenied(): void {
+		$this->createEnvironment(
+			[],
+			function ( Client $client, FunFunFactory $factory ): void {
+				$this->setDefaultSkin( $factory, 'cat17' );
+				$donation = $this->newStoredDonation( $factory );
+				$donorData = $this->newPrivateDonorData();
+				$donorData['email'] = 'this_is_not_a_valid_email_address.de';
+				$crawler = $this->performRequest(
+					$client,
+					$donorData,
+					$donation->getId(),
+					self::CORRECT_UPDATE_TOKEN,
+					self::CORRECT_UPDATE_TOKEN
+				);
+
+				$response = $client->getResponse();
+				$this->assertTrue( $response->isSuccessful() );
+				$this->assertContains(
+					'donor_change_failure_validation_error',
+					$crawler->filter( '.messages-donor' )->html()
+				);
+			}
+		);
+	}
+
+	public function testWhenDonationAlreadyHasAddress_requestIsDenied(): void {
+		$this->createEnvironment(
+			[],
+			function ( Client $client, FunFunFactory $factory ): void {
+				$this->setDefaultSkin( $factory, 'cat17' );
+
+				$donation = ValidDoctrineDonation::newDirectDebitDoctrineDonation();
+				$donation->modifyDataObject(
+					function ( DonationData $data ) {
+						$data->setAccessToken( self::CORRECT_UPDATE_TOKEN );
+						$data->setUpdateToken( self::CORRECT_UPDATE_TOKEN );
+					}
+				);
+				$factory->getEntityManager()->persist( $donation );
+				$factory->getEntityManager()->flush();
+
+				$this->performRequest(
+					$client,
+					$this->newPrivateDonorData(),
+					$donation->getId(),
+					self::CORRECT_UPDATE_TOKEN,
+					self::CORRECT_UPDATE_TOKEN
+				);
+				$response = $client->getResponse();
+				$this->assertTrue( $response->isForbidden() );
+			}
+		);
+	}
+
+	private function performRequest( Client $client, array $data, int $donationId, string $accessToken, string $updateToken ): Crawler {
+		return $client->request(
+			Request::METHOD_POST,
+			self::PATH . '?accessToken=' . $accessToken,
+			array_merge(
+				[
+					'donation_id' => $donationId,
+					'updateToken' => $updateToken
+				],
+				$data
+			)
+		);
+	}
+
+	private function newStoredDonation( FunFunFactory $factory ): Donation {
+		$factory->setDonationTokenGenerator(
+			new FixedTokenGenerator(
+				self::CORRECT_UPDATE_TOKEN,
+				new \DateTime( '9001-01-01' )
+			)
+		);
+
+		$donation = ValidDonation::newIncompleteAnonymousPayPalDonation();
+
+		$factory->getDonationRepository()->storeDonation( $donation );
+
+		return $donation;
+	}
+
+	private function newPrivateDonorData(): array {
+		return [
+			'addressType' => DonorName::PERSON_PRIVATE,
+			'city' => 'Mönchengladbach',
+			'companyName' => '',
```

Comment:
You should leave out the empty fields, the browsers won't send them and our application should be robust enough to initialize the variables with empty strings.